x/mongo/driver/topology/pool_test.go

```diff
@@ -175,6 +175,52 @@ func TestPool(t *testing.T) {
 				t.Errorf("Should have set the connection state on return. got %d; want %d", state, disconnected)
 			}
 		})
+		t.Run("no race if connections are also connecting", func(t *testing.T) {
+			cleanup := make(chan struct{})
+			addr := bootstrapConnections(t, 3, func(nc net.Conn) {
+				<-cleanup
+				_ = nc.Close()
+			})
+			d := newdialer(&net.Dialer{})
+			pc := poolConfig{
+				Address: address.Address(addr.String()),
+			}
+			p, err := newPool(pc, WithDialer(func(Dialer) Dialer { return d }))
+			noerr(t, err)
+			err = p.connect()
+			noerr(t, err)
+			getDone := make(chan struct{})
+			disconnectDone := make(chan struct{})
+			_, err = p.get(context.Background())
+			noerr(t, err)
+			getCtx, getCancel := context.WithCancel(context.Background())
+			defer getCancel()
+			go func() {
+				defer close(getDone)
+				for {
+					select {
+					case <-disconnectDone:
+						return
+					default:
+						_, _ = p.get(getCtx)
+						noerr(t, err)
+						time.Sleep(time.Microsecond)
+					}
+				}
+			}()
+			go func() {
+				_, err := p.get(getCtx)
+				noerr(t, err)
+				ctx, cancel := context.WithTimeout(context.Background(), 3*time.Microsecond)
```

Comment:
Did this allow you to trigger the race with -race? Generally, you have to not use time and instead use channels to coordinate "steps". For instance, you can have each goroutine get into the right state by closing a series of channels in order to force the race, if possible. Without the right internal instrumentation though, you may just need to show that with -count=N, the race happens.

Reply:
Do you have an example of this? This triggers the race once every few runs, but I'd be happy to switch it to something less hack-y and more consistent.

Reply:
yes! https://github.com/edaniels/go-net-utils/blob/master/track/dialer_test.go#L143